samples/sample-07/src/test/java/com/example/sample07/Sample07ApplicationTest.java

```diff
@@ -0,0 +1,335 @@
+		props.put(ConsumerConfig.GROUP_PROTOCOL_CONFIG, "consumer");
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		return props;
+	}
+
+	static Map<String, String> getPropertiesFromFile() {
+		final Resource resource = new ClassPathResource(BROKER_PROPERTIES_FILE_PATH);
+		try {
+			final Properties properties = PropertiesLoaderUtils.loadAllProperties(resource.getFilename());
+			return properties.stringPropertyNames().stream()
+							.collect(Collectors.toMap(
+										s -> s,
+										s -> (String) properties.get(s)));
+		}
+		catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@EnableKafka
+	@TestConfiguration
+	public static class TestConfig {
+
+		CountDownLatch partitionRevokedLatch = new CountDownLatch(0);
+
+		CountDownLatch partitionAssignedLatch = new CountDownLatch(1);
+
+		CountDownLatch listenerLatch = new CountDownLatch(2);
+
+		@KafkaListener(id = GROUP_ID, topics = TOPIC_NAME, autoStartup = "false")
+		void listen(ConsumerRecord<Integer, String> ignored) {
+			listenerLatch.countDown();
+		}
+
+		@Bean
+		KafkaTemplate<Integer, String> template(ProducerFactory<Integer, String> producerFactory) {
+			return new KafkaTemplate<>(producerFactory);
+		}
+
+		@Bean
+		ProducerFactory<Integer, String> producerFactory() {
+			return new DefaultKafkaProducerFactory<>(KafkaTestUtils.producerProps(BROKER));
+		}
+
+		@Bean
+		ConcurrentKafkaListenerContainerFactory<Integer, String>
+		kafkaListenerContainerFactory(ConsumerFactory<Integer, String> consumerFactory) {
+
+			final ConcurrentKafkaListenerContainerFactory<Integer, String> factory =
+					new ConcurrentKafkaListenerContainerFactory<>();
+
+			factory.setConsumerFactory(consumerFactory);
+
+			factory.getContainerProperties().setConsumerRebalanceListener(new ConsumerAwareRebalanceListener() {
+				@Override
+				public void onPartitionsAssigned(Consumer<?, ?> consumer,
+												Collection<TopicPartition> partitions) {
+					partitionAssignedLatch.countDown();
+				}
+
+				@Override
+				public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+					partitionRevokedLatch.countDown();
+				}
+			});
+			return factory;
+		}
+
+		@Bean
+		ConsumerFactory<Integer, String> consumerFactory() throws InterruptedException {
+			Properties propsDummy = new Properties();
+			propsDummy.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class.getName());
+			propsDummy.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+									StringDeserializer.class.getName());
+			propsDummy.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER);
+			propsDummy.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "hello");
+			propsDummy.setProperty(ConsumerConfig.GROUP_PROTOCOL_CONFIG, "consumer");
+			propsDummy.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+			Consumer<Object, Object> dummyConsumer = new KafkaConsumer<>(propsDummy);
+
+			Thread.sleep(5000);
+			dummyConsumer.close();
```

Comment:
This looks like a mix of concerns.
Why cannot we just rely on the auto-configuration for the `ConsumerFactory` and then have such a dummy consumer created from the `ApplicationRunner`?

Reply:
> I guess that new consumer protocol has tiny concurrent issue. (I reported it via https://issues.apache.org/jira/browse/KAFKA-16670)
> In case of 3, my work around is using dummy consumer to make broker load metadata from __consumer_offsets.

I don't think so. it is intended. 😄 
As i said in PR description, this is work around to solve tiny issue. 

1. Right after the `broker` starts, `broker` seems not load metadata from `__consumer_offsets`. Once the metadata is loaded, the `broker` can function as a normal coordinator.
2. When a `consumer` is created, `consumer` sends a `findCoordinator` request to the `bootstrap server`.
3. Upon receiving `findCoordinator` request, the `broker` schedules a metadata load from `__consumer_offsets`. However, at this point, since the `broker` has not loaded the metadata yet.

and then, `consumer.subscribe()` and `consumer.poll()` will be executed. 
If `broker` does not load metadata yet, `consumer` received response of first `consumer.poll() and `Response` said `this coordinator is invalid`. 
And then, `consumer` try to find valid `coordinator`, but stuck in that status forever. 

This is problem which what i said. 
To solve this problem, I create `dummyConsumer` to make broker schedule loading metadata.  (and `dummyConsumer` will be closed after sleeping 5 sec, because it's only needed to trigger schedule). 

This is why i make other `consumer` without `ConsumerFactory`! 🤔 
What do you think?

Reply:
> And then, consumertry to find validcoordinator`, but stuck in that status forever.

If that is the problem of Apache Kafka by itself for this new algorithm, how about to postpone such a sample until it is solved?
I wonder how end-user are supposed to solve this problem in the applications?
If you still insist having this sample, how about to rework it the way end-user would implement it in their own applications?
No one is going to do test stuff in production.
The sample purpose is to get it is and have your own application up-and-running.

Reply:
Um... I see! 
I will try to find common way.  
If i fail to find a good common way, it might be better to close this PR. 🤔